### PR TITLE
Upgrade all packages using `pur` tool

### LINKED ISSUE #1893: Bump allauth to next release
When the next version of allauth is released, be sure to bump allauth in requirements.

### PR DESCRIPTION
**Pull request by humitos**
These ones were not upgraded since they are incompatible:

* django
* docker-py
* celery
* elasticsearch
* pyelasticsearch

This is just a way to run all the test with this environment and also to discuss with other people if this is something we really want.

Closes #1893

### CODE REVIEW

**Comment by humitos** [comment]
@agjohnson is this something that we want or do we prefer to upgrade packages one by one? Let me know, so I can update the PR or just close it.

**Comment by humitos** [comment]
Updated today to the latest versions compatible with our codebase.

@ericholscher @agjohnson please, let me know if you are interested on these kind of upgrade or not.

- I didn't touch `docker-py` here because it's done at #3243 
- `django` needs extra work, which is being done at #3372 
- `elasticsearch` didn't touch since Eric is working on that #3373

**Comment by ericholscher** [comment]
I think it's a good practice. I will let this sit until Friday then merge it, so we can do some local testing etc on it before deploy.

**Comment by humitos** [comment]
It's Friday night here, nobody is around and Xmas is comming this weekend. So, maybe it's a better idea to merge this next week :)

File: requirements/pip.txt [hunk 1]
```diff
@@ -1,73 +1,86 @@
...
-Sphinx==1.5.3
+virtualenv==15.1.0
+docutils==0.14
+Sphinx==1.6.5
 sphinx_rtd_theme==0.2.5b1
 Pygments==2.2.0
-mkdocs==0.14.0
-django==1.9.12
-six==1.10.0
+mkdocs==0.17.2
```
[the remark this answers is not in the document]
Comment (humitos) [reply requirements/pip.txt]: Why not? Do we you have an example where it fails?

I could write a test to make it fail so we don't hit this issue again and downgrade to `0.14.0` as it was before this PR.

**Comment by humitos** [comment]
@agjohnson I updated this PR with the latest packages

**Comment by humitos** [comment]
I didn't touch the `deploy.txt` requeriments, but just in case we want to upgrade it, here is the output:

```
$ pur --skip django-tastypie,django,docker-py,elasticsearch,pyelasticsearch,commonmark,stripe,djangorestframework -r requirements/deploy.txt 
New version for readthedocs-build found (2.0.9), but current spec prohibits updating: readthedocs-build<2.1
Updated gunicorn: 19.1.0 -> 19.7.1
Updated pysolr: 2.0.13 -> 3.7.0
Updated django-redis-cache: 1.6.3 -> 1.7.1
Updated django-mailgun: 0.2.2 -> 0.9.1
Updated hiredis: 0.1.2 -> 0.2.0
Updated cssselect: 0.7.1 -> 1.0.3
All requirements up-to-date.
```

Should I commit this changes? Is it safe?

**Comment by xrmx** [comment]
You may want to upgrade to `linaro_django_pagination==2.0.3` as seen on https://github.com/rtfd/readthedocs.org/pull/3305 too

**Comment by humitos** [comment]
@xrmx I didn't upgrade it because it pinned to a very specific commit from a fork of the original repo (I tried upgrading it and test failed) --and as I don't know the main reason I prefered to leave it as it was.

In fact, in the issue you linked here, they are talking about some problems with that package and they don't know how to fix it.

So, I wouldn't touch it without knowing the implicances. Maybe, as they are saying is good to remove that dependency after a research on how it would affect.

**Comment by humitos** [comment]
Tow more related PRs that I wanted to include in this branch:

* upgrade django pagination #3666 
* upgrade django taggit #3667

**Comment by humitos** [comment]
@xrmx maybe you are interested in my last two PRs from my last comment :)

**Comment by xrmx** [comment]
@humitos nice, thanks!

**Comment by xrmx** [comment]
Wondering if pyelasticsearch can be dropped, git grep pyelasticsearch returns nothing and (current) haystack uses elasticsearch-py.

**Comment by humitos** [comment]
@xrmx maybe. How elasticsearch works currently in production is a mistery for me, so I didn't want to get involved on that. As I don't have experience with that I didn't touch it.

**Comment by xrmx** [comment]
@humitos if nobody uses it, what can it break? Anyway i've removed it from my local installation and nothing changed.

**Comment by humitos** [comment]
@xrmx I'm not saying that nobody uses it. I'm saying that _I don't know how it's used, if it is_ and that's why I didn't wanted to remove it.

The search engine is complex (in fact, you opened another issue saying that it doesn't work your local instance). So, you don't know if it doesn't break things.

File: requirements/pip.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (agjohnson) [reply requirements/pip.txt]: mkdocs `0.17` doesn't work at all with RTD. They removed a feature we required. `0.16.x` was the last working version

**Comment by agjohnson** [comment]
elasticsearch is indeed used, it shouldn't be removed.

This has a high breakage potential. We should plan on merging on a week where we have plenty of eyes on the deploy. I don't have any strong feedback here otherwise, so effectively +1. Perhaps next week we can merge early, QA, then deploy midweek?

**Comment by xrmx** [comment]
@agjohnson we were talking about pyelasticsearch not elasticsearch :)

**Comment by ericholscher** [comment]
@humitos we have mkdocs there so that people running it locally can build mkdocs docs. This is less important now that we have docker support. There is likely a lot that could be removed that is a build dependency, but not required for basic dev.

**Comment by humitos** [comment]
@ericholscher thanks for the explanation about this. Considering what you have said, I think it's a good idea to split our `requirements/pip.txt` to have all the RTD specific requirements and another one for local builds (using `LocalBuildEnvironment` class). So, if you are running Docker you don't need to install them.

Besides, it easier to know what dependency is for. What do you think?

I will downgrade `mkdocs` to keep it compatible as @agjohnson said and if you like my proposal, I will achieve that in another PR.

File: requirements/pip.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (humitos) [reply requirements/pip.txt]: I pinned to 0.15.0 --the same that we install by default in the venvs

**Comment by ericholscher** [comment]
Big +1 on simplifying the requirements file for local dev. It should def. be in another PR, but I think it is a big reason that people have a hard time getting started, so reducing the "time to running local instance" is a good metric to increase contribution!

**Comment by agjohnson** [comment]
Rebased this, should hopefully be able to be merged after tests.

**Comment by humitos** [comment]
I think we are all good now.

Some context: 

* `mock==1.0` probably wasn't checking that the `assert_` method exist in the mock object and was considering it as another mocked method (arbitrary) that you can call --that's why it didn't fail in the test.
* now, `mock==2.0` is checking this and the test start failing because the name with the final `s` was a mistake
* finally, there was a bug on how the `broadcast` method was being called inside the `Project.save` since it was using the `ProjectRelationship.pk` instead of the proper `pk` from the superproject :)

**Comment by agjohnson** [comment]
Nice! I wonder if this actually resolves some of the problems with symlinking that we noticed in the deploys.

I'll go ahead and merge this and the docker upgrade as well.

File: requirements/pip.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidfischer) [reply requirements/pip.txt]: Do you know which feature was removed in 0.17? There's at least one ticket (https://github.com/rtfd/readthedocs.org/issues/3991) where the recommended fix from an mkdocs contributor was to upgrade.

**Pull request merged**